[Rust] Cannot build with KUZU_SHARED

### Kùzu version

v0.4.2

### What happened?

When building kuzu 0.4.2 with cargo using with gcc 11 for nix-os (linux)

`KUZU_SHARED=true`
`KUZU_INCLUDE_LIB=/home/dimpdash/code/context-io/build-kuzu`
`KUZU_LIBRARY_LIB=/home/dimpdash/code/context-io/build-kuzu`

downloaded prebuilts from
`curl -L -O https://github.com/kuzudb/kuzu/releases/download/v0.4.2/libkuzu-linux-x86_64.tar.gz `


I get the build error 

```
error: failed to run custom build command for `kuzu v0.4.2`

Caused by:
  process didn't exit successfully: `/home/dimpdash/code/context-io/target/debug/build/kuzu-cb2e966705bb8007/build-script-build` (exit status: 1)
  --- stdout
  cargo:rustc-link-search=native=/home/dimpdash/code/context-io/build-kuzu
  cargo:rustc-link-lib=dylib=kuzu
  cargo:CXXBRIDGE_PREFIX=kuzu
  cargo:CXXBRIDGE_LINKS=kuzu
  cargo:CXXBRIDGE_DIR0=/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/include
  cargo:CXXBRIDGE_DIR1=/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/crate
  cargo:rustc-link-lib=dylib=kuzu
  cargo:rerun-if-env-changed=KUZU_SHARED
  cargo:rerun-if-changed=include/kuzu_rs.h
  cargo:rerun-if-changed=include/kuzu_rs.cpp
  TARGET = Some("x86_64-unknown-linux-gnu")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-unknown-linux-gnu")
  cargo:rerun-if-env-changed=CXX_x86_64-unknown-linux-gnu
  CXX_x86_64-unknown-linux-gnu = None
  cargo:rerun-if-env-changed=CXX_x86_64_unknown_linux_gnu
  CXX_x86_64_unknown_linux_gnu = None
  cargo:rerun-if-env-changed=HOST_CXX
  HOST_CXX = None
  cargo:rerun-if-env-changed=CXX
  CXX = Some("/nix/store/3n2d67nqxinplhd841iw23369pdsgnqm-gcc-wrapper-11.4.0/bin/g++")
  cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  cargo:rerun-if-env-changed=CXXFLAGS_x86_64-unknown-linux-gnu
  CXXFLAGS_x86_64-unknown-linux-gnu = None
  cargo:rerun-if-env-changed=CXXFLAGS_x86_64_unknown_linux_gnu
  CXXFLAGS_x86_64_unknown_linux_gnu = None
  cargo:rerun-if-env-changed=HOST_CXXFLAGS
  HOST_CXXFLAGS = None
  cargo:rerun-if-env-changed=CXXFLAGS
  CXXFLAGS = None
  cargo:warning=In file included from /nix/store/vrccdh900iy0331y4mam2gpwmjvys3sx-gcc-11.4.0/include/c++/11.4.0/x86_64-unknown-linux-gnu/bits/os_defines.h:39,
  cargo:warning=                 from /nix/store/vrccdh900iy0331y4mam2gpwmjvys3sx-gcc-11.4.0/include/c++/11.4.0/x86_64-unknown-linux-gnu/bits/c++config.h:586,
  cargo:warning=                 from /nix/store/vrccdh900iy0331y4mam2gpwmjvys3sx-gcc-11.4.0/include/c++/11.4.0/bits/stl_algobase.h:59,
  cargo:warning=                 from /nix/store/vrccdh900iy0331y4mam2gpwmjvys3sx-gcc-11.4.0/include/c++/11.4.0/memory:63,
  cargo:warning=                 from /home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/crate/kuzu/include/kuzu_rs.h:3,
  cargo:warning=                 from /home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/sources/kuzu/src/ffi.rs.cc:1:
  cargo:warning=/nix/store/s3pvsv4as7mc8i2nwnk2hnsyi2qdj4bq-glibc-2.39-31-dev/include/features.h:414:4: warning: #warning _FORTIFY_SOURCE requires compiling with optimization (-O) [-Wcpp]
  cargo:warning=  414 | #  warning _FORTIFY_SOURCE requires compiling with optimization (-O)
  cargo:warning=      |    ^~~~~~~
  cargo:warning=In file included from /nix/store/vrccdh900iy0331y4mam2gpwmjvys3sx-gcc-11.4.0/include/c++/11.4.0/x86_64-unknown-linux-gnu/bits/os_defines.h:39,
  cargo:warning=                 from /nix/store/vrccdh900iy0331y4mam2gpwmjvys3sx-gcc-11.4.0/include/c++/11.4.0/x86_64-unknown-linux-gnu/bits/c++config.h:586,
  cargo:warning=                 from /nix/store/vrccdh900iy0331y4mam2gpwmjvys3sx-gcc-11.4.0/include/c++/11.4.0/bits/stl_algobase.h:59,
  cargo:warning=                 from /nix/store/vrccdh900iy0331y4mam2gpwmjvys3sx-gcc-11.4.0/include/c++/11.4.0/memory:63,
  cargo:warning=                 from /home/dimpdash/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/include/kuzu_rs.h:3,
  cargo:warning=                 from src/kuzu_rs.cpp:1:
  cargo:warning=/nix/store/s3pvsv4as7mc8i2nwnk2hnsyi2qdj4bq-glibc-2.39-31-dev/include/features.h:414:4: warning: #warning _FORTIFY_SOURCE requires compiling with optimization (-O) [-Wcpp]
  cargo:warning=  414 | #  warning _FORTIFY_SOURCE requires compiling with optimization (-O)
  cargo:warning=      |    ^~~~~~~
  cargo:warning=In file included from /home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/sources/kuzu/src/ffi.rs.cc:1:
  cargo:warning=/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/crate/kuzu/include/kuzu_rs.h:213:23: error: ‘storage_version_t’ in namespace ‘kuzu::storage’ does not name a type
  cargo:warning=  213 | inline kuzu::storage::storage_version_t get_storage_version() {
  cargo:warning=      |                       ^~~~~~~~~~~~~~~~~
  cargo:warning=/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/sources/kuzu/src/ffi.rs.cc: In function ‘uint64_t kuzu_rs::kuzu_rs$cxxbridge1$get_storage_version()’:
  cargo:warning=/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/sources/kuzu/src/ffi.rs.cc:1581:58: error: ‘get_storage_version’ is not a member of ‘kuzu_rs’
  cargo:warning= 1581 |   ::std::uint64_t (*get_storage_version$)() = ::kuzu_rs::get_storage_version;
  cargo:warning=      |                                                          ^~~~~~~~~~~~~~~~~~~
  cargo:warning=In file included from src/kuzu_rs.cpp:1:
  cargo:warning=/home/dimpdash/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/include/kuzu_rs.h:213:23: error: ‘storage_version_t’ in namespace ‘kuzu::storage’ does not name a type
  cargo:warning=  213 | inline kuzu::storage::storage_version_t get_storage_version() {
  cargo:warning=      |                       ^~~~~~~~~~~~~~~~~
  exit status: 1
  cargo:warning=ToolExecError: Command "/nix/store/3n2d67nqxinplhd841iw23369pdsgnqm-gcc-wrapper-11.4.0/bin/g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-m64" "-I" "/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/include" "-I" "/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/crate" "-I" "/home/dimpdash/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/include" "-I" "/home/dimpdash/code/context-io/build-kuzu" "-Wall" "-Wextra" "-std=c++2a" "-DENABLE_RUNTIME_CHECKS=1" "-o" "/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/0d1e03e935d2d86a-ffi.rs.o" "-c" "/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/sources/kuzu/src/ffi.rs.cc" with args g++ did not execute successfully (status code exit status: 1).
  exit status: 1
  cargo:warning=ToolExecError: Command "/nix/store/3n2d67nqxinplhd841iw23369pdsgnqm-gcc-wrapper-11.4.0/bin/g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-m64" "-I" "/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/include" "-I" "/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/crate" "-I" "/home/dimpdash/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/include" "-I" "/home/dimpdash/code/context-io/build-kuzu" "-Wall" "-Wextra" "-std=c++2a" "-DENABLE_RUNTIME_CHECKS=1" "-o" "/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/2e40c9e35e9506f4-kuzu_rs.o" "-c" "src/kuzu_rs.cpp" with args g++ did not execute successfully (status code exit status: 1).

  --- stderr

  CXX include path:
    /home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/include
    /home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/crate


  error occurred: Command "/nix/store/3n2d67nqxinplhd841iw23369pdsgnqm-gcc-wrapper-11.4.0/bin/g++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-gdwarf-4" "-fno-omit-frame-pointer" "-m64" "-I" "/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/include" "-I" "/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/cxxbridge/crate" "-I" "/home/dimpdash/.cargo/registry/src/index.crates.io-6f17d22bba15001f/kuzu-0.4.2/include" "-I" "/home/dimpdash/code/context-io/build-kuzu" "-Wall" "-Wextra" "-std=c++2a" "-DENABLE_RUNTIME_CHECKS=1" "-o" "/home/dimpdash/code/context-io/target/debug/build/kuzu-efed61c5fa8de02a/out/2e40c9e35e9506f4-kuzu_rs.o" "-c" "src/kuzu_rs.cpp" with args g++ did not execute successfully (status code exit status: 1).

```

### Are there known steps to reproduce?

_No response_

--- a/.github/workflows/ci-workflow.yml
+++ b/.github/workflows/ci-workflow.yml
@@ -71,6 +71,14 @@ jobs:
       - name: Java test
         run: make javatest
 
+      - name: Rust test with pre-built library
+        working-directory: tools/rust_api
+        env:
+          KUZU_SHARED: 1
+          KUZU_INCLUDE_DIR: ${{ github.workspace }}/build/release/src
+          KUZU_LIBRARY_DIR: ${{ github.workspace }}/build/release/src
+        run: cargo test --locked --features arrow -- --test-threads=1
+
   code-coverage:
     name: code coverage
     runs-on: ubuntu-22.04
@@ -217,6 +225,14 @@ jobs:
       - name: Java test
         run: make javatest
 
+      - name: Rust test with pre-built library
+        working-directory: tools/rust_api
+        env:
+          KUZU_SHARED: 1
+          KUZU_INCLUDE_DIR: ${{ github.workspace }}/build/release/src
+          KUZU_LIBRARY_DIR: ${{ github.workspace }}/build/release/src
+        run: cargo test --locked --features arrow -- --test-threads=1
+
   msvc-build-test:
     name: msvc build & test
     needs: [clang-format, sanity-checks, python-lint-check]
@@ -278,7 +294,6 @@ jobs:
         working-directory: tools/rust_api
         run: |
           call "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvars64.bat"
-          set KUZU_TESTING=1
           set CFLAGS=/MD
           set CXXFLAGS=/MD /std:c++20
           cargo test --locked -- --test-threads=1
@@ -288,6 +303,19 @@ jobs:
         run: |
           call "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvars64.bat"
           make javatest
+
+      - name: Rust test with pre-built library
+        working-directory: tools/rust_api
+        env:
+          KUZU_SHARED: 1
+          KUZU_INCLUDE_DIR: ${{ github.workspace }}/build/release/src
+          KUZU_LIBRARY_DIR: ${{ github.workspace }}/build/release/src
+        shell: cmd
+        run: |
+          call "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvars64.bat"
+          set PATH=%PATH%;${{ github.workspace }}/build/release/src
+          cargo test --locked --features arrow -- --test-threads=1
+
 
   sanity-checks:
     name: sanity checks
@@ -474,6 +502,10 @@ jobs:
           ulimit -n 10240
           source /Users/runner/.cargo/env
           cargo build --locked --features arrow
+
+      # TODO(bmwinger): rust pre-built library test for macos
+      # When last tried it had issues finding libkuzu.dylib when running the tests
+      # Using LD_LIBRARY_PATH (and similar variables) did not work for some reason
 
   shell-test:
     name: shell test

--- a/Makefile
+++ b/Makefile
@@ -121,7 +121,6 @@ python-debug:
 
 rust:
 ifeq ($(OS),Windows_NT)
-	set KUZU_TESTING=1
 	set CFLAGS=/MDd
 	set CXXFLAGS=/MDd /std:c++20
 	set CARGO_BUILD_JOBS=$(NUM_THREADS)

--- a/src/include/common/types/types.h
+++ b/src/include/common/types/types.h
@@ -425,7 +425,7 @@ private:
     uint64_t numElements;
 };
 
-class StructField {
+class KUZU_API StructField {
 public:
     StructField() : type{std::make_unique<LogicalType>()} {}
     StructField(std::string name, std::unique_ptr<LogicalType> type)
@@ -485,7 +485,7 @@ private:
 
 using logical_type_vec_t = std::vector<LogicalType>;
 
-struct ListType {
+struct KUZU_API ListType {
     static LogicalType getChildType(const LogicalType& type);
 };
 
@@ -494,7 +494,7 @@ struct KUZU_API ArrayType {
     static uint64_t getNumElements(const LogicalType& type);
 };
 
-struct StructType {
+struct KUZU_API StructType {
     static std::vector<LogicalType> getFieldTypes(const LogicalType& type);
 
     static std::vector<std::string> getFieldNames(const LogicalType& type);
@@ -512,13 +512,13 @@ struct StructType {
     static struct_field_idx_t getFieldIdx(const LogicalType& type, const std::string& key);
 };
 
-struct MapType {
+struct KUZU_API MapType {
     static LogicalType getKeyType(const LogicalType& type);
 
     static LogicalType getValueType(const LogicalType& type);
 };
 
-struct UnionType {
+struct KUZU_API UnionType {
     static constexpr union_field_idx_t TAG_FIELD_IDX = 0;
 
     static constexpr LogicalTypeID TAG_FIELD_TYPE = LogicalTypeID::INT8;

--- a/src/include/main/kuzu.h
+++ b/src/include/main/kuzu.h
@@ -20,3 +20,4 @@
 #include "main/storage_driver.h"              // IWYU pragma: export
 #include "main/version.h"                     // IWYU pragma: export
 #include "processor/result/flat_tuple.h"      // IWYU pragma: export
+#include "storage/storage_version_info.h"     // IWYU pragma: export

--- a/src/include/storage/storage_version_info.h
+++ b/src/include/storage/storage_version_info.h
@@ -4,6 +4,8 @@
 #include <string>
 #include <unordered_map>
 
+#include "common/api.h"
+
 namespace kuzu {
 namespace storage {
 
@@ -18,7 +20,7 @@ struct StorageVersionInfo {
             {"0.0.3", 1}};
     }
 
-    static storage_version_t getStorageVersion();
+    static KUZU_API storage_version_t getStorageVersion();
 
     static constexpr const char* MAGIC_BYTES = "KUZU";
 };

--- a/tools/rust_api/Cargo.toml
+++ b/tools/rust_api/Cargo.toml
@@ -22,7 +22,6 @@ include = [
 ]
 
 edition = "2018"
-links = "kuzu"
 
 [dependencies]
 cxx = "1.0"

--- a/tools/rust_api/build.rs
+++ b/tools/rust_api/build.rs
@@ -18,7 +18,11 @@ fn get_target() -> String {
 }
 
 fn link_libraries() {
-    println!("cargo:rustc-link-lib={}=kuzu", link_mode());
+    if cfg!(windows) && link_mode() == "dylib" {
+        println!("cargo:rustc-link-lib=dylib=kuzu_shared");
+    } else {
+        println!("cargo:rustc-link-lib={}=kuzu", link_mode());
+    }
     if link_mode() == "static" {
         if cfg!(windows) {
             println!("cargo:rustc-link-lib=dylib=msvcrt");
@@ -132,8 +136,6 @@ fn build_ffi(
 
     build.includes(include_paths);
 
-    link_libraries();
-
     println!("cargo:rerun-if-env-changed=KUZU_SHARED");
 
     println!("cargo:rerun-if-changed=include/kuzu_rs.h");
@@ -162,15 +164,15 @@ fn main() {
         (env::var("KUZU_LIBRARY_DIR"), env::var("KUZU_INCLUDE_DIR"))
     {
         println!("cargo:rustc-link-search=native={}", kuzu_lib_dir);
-        if cfg!(windows) && link_mode() == "dylib" {
-            println!("cargo:rustc-link-lib=dylib=kuzu_shared");
-        } else {
-            println!("cargo:rustc-link-lib={}=kuzu", link_mode());
-        }
+        // Also add to environment used by cargo run and tests
+        println!("cargo:rustc-env=LD_LIBRARY_PATH={}", kuzu_lib_dir);
         include_paths.push(Path::new(&kuzu_include).to_path_buf());
     } else {
         include_paths.extend(build_bundled_cmake().expect("Bundled build failed!"));
         bundled = true;
+    }
+    if link_mode() == "static" {
+        link_libraries();
     }
     build_ffi(
         "src/ffi.rs",
@@ -188,5 +190,8 @@ fn main() {
             bundled,
             &include_paths,
         );
+    }
+    if link_mode() == "dylib" {
+        link_libraries();
     }
 }

--- a/tools/rust_api/src/database.rs
+++ b/tools/rust_api/src/database.rs
@@ -1,4 +1,4 @@
-use cxx::{let_cxx_string, UniquePtr};
+use cxx::UniquePtr;
 use std::cell::UnsafeCell;
 use std::fmt;
 use std::path::Path;
@@ -115,9 +115,10 @@ mod tests {
     // working with multiple databases in parallel.
     // Tests can be run serially with `cargo test -- --test-threads=1` to work around this.
 
+    #[test]
     fn create_database() -> Result<()> {
         let temp_dir = tempfile::tempdir()?;
-        let mut db = Database::new(temp_dir.path(), SystemConfig::default())?;
+        let _db = Database::new(temp_dir.path(), SystemConfig::default())?;
         temp_dir.close()?;
         Ok(())
     }

--- a/tools/rust_api/src/query_result.rs
+++ b/tools/rust_api/src/query_result.rs
@@ -202,7 +202,6 @@ mod tests {
     use crate::connection::Connection;
     use crate::database::{Database, SystemConfig};
     use crate::logical_type::LogicalType;
-    use crate::query_result::CSVOptions;
 
     #[test]
     fn test_query_result_metadata() -> anyhow::Result<()> {

